New: Add `overrides` and `files` configuration options (refs #3611)

smably opened the pull request:
**What is the purpose of this pull request? (put an "X" next to item)**

- [x] Documentation update
- [x] Other, please explain:


The ability to override config values based on glob pattern matching relative to the config path. #3611

**What changes did you make? (Give an overview)**

Starting from @CrabDude's work in #7177, I rebased the branch against the latest master, addressed several comments from the original PR, rewrote and expanded many of the JSDoc comments, tweaked documentation, and split as much logic as I could from `lib/config.js` into `lib/config/config-cache.js`, `lib/config/config-file.js`, and `lib/config/config-ops.js`.


**Is there anything you'd like reviewers to focus on?**

Validating the approach -- does the separation of logic make sense? Is this still on the right track? Any edge cases that need tests?

not-an-aardvark commented:
Thanks! Sorry, looks like we lost track of this PR.

To make this easier to review, how difficult would it be to split up @CrabDude's commit and your improvements? It's currently a bit difficult to identify which parts are different from https://github.com/eslint/eslint/pull/7177.

Also, does this implement the changes that @nzakas requested in https://github.com/eslint/eslint/pull/7177#pullrequestreview-1184437?

smably commented:
Hi @not-an-aardvark, no worries. Unfortunately it would be pretty difficult to split up, because I had to rewrite large parts of @CrabDude's work to ES6-ify it (the diff was pretty much useless because the indentation changed when I moved everything inside a class). I can see about trying to split it up, but I think it might be best to review this from scratch as the changes are pretty substantial (particularly when it comes to the caches). Let me know what you think.

I'll add some comments on @nzakas' review on the original PR. Sorry I couldn't maintain the original PR -- not sure if there's a way to do that on a PR from a personal fork.

If there is still potential for this to get merged, I'll definitely see if I can rebase and fix the conflict later today.

smably commented:
Added comments to the original PR and fixed the merge conflict.

smably commented:
@not-an-aardvark Turned out I had I still had the branch with the unsquashed version of this locally. I've updated this PR with the contents of the unsquashed branch. The first commit is the existing work, with some updates to make it mergeable but no functional changes. The subsequent commits represent the new stuff I've changed or added.

not-an-aardvark commented:
Thanks!

I'm planning to review this sometime in the next day or two. I'm not very familiar with the part of the codebase that this changes, so I'm hoping someone else can also review it.

ilyavolodin commented on lib/config/config-ops.js:
```diff
@@ -268,5 +272,84 @@ module.exports = {
+     */
+    pathMatchesGlobs(filePath, patterns) {
+        patterns = [].concat(patterns);
+        return patterns.every(pattern => minimatch(filePath, pattern, { matchBase: true }));
```
We are not protecting against "../" patterns anywhere. Those will basically never work (or at least will not work in most cases). They are also going to be really hard to wrap your head around, so I think we should filter out and ignore them.

ilyavolodin replied:
Also, I'm slightly concerned about absolute paths as well.

smably replied:
Fixed both (will now ignore patterns with absolute paths or that contain `..`).

ilyavolodin commented on lib/config/config-cache.js:
```diff
@@ -0,0 +1,150 @@
+/**
```
We now have two config caches. One for config files themselves (used in file-finder.js) and one for config vectors. Do we need both of them?

ilyavolodin commented on lib/config/config-ops.js:
```diff
@@ -268,5 +272,84 @@ module.exports = {
+            subvector.pop();
+        }
+
+        if (config) {
```
This is not a very helpful comment, but all of the code that goes below seems like a black magic to me. Branching on array value types, hardcoded array indexes, etc. I sort of understand what the code is doing, but it's not very readable and is really hard to follow. Not sure what can be done to improve this (as I said, not a very helpful comment).

not-an-aardvark replied:
Agreed. Using array methods like `.find()` instead of loops would probably be a good idea.

not-an-aardvark replied:
Using a more organized object instead of a "vector" with multiple datatypes would also make this easier to read.

smably replied:
Fixed. I refactored all the code relating to vectors (still can't think of a better name) and it is now vastly more readable at the expense of a bit of performance.

ilyavolodin commented:
Performance numbers look pretty good (although we are not testing for anything with overrides):

Before:

```
Single File:
  CPU Speed is 3500 with multiplier 13000000
  Performance Run #1:  5168.491679ms
  Performance Run #2:  3631.0344720000003ms
  Performance Run #3:  3644.98604ms
  Performance Run #4:  3630.521594ms
  Performance Run #5:  3626.206455ms

  Performance budget ok:  3631.0344720000003ms (limit: 3714.285714285714ms)


Multi Files (0 files):
  CPU Speed is 3500 with multiplier 39000000
  Performance Run #1:  13425.494453ms
  Performance Run #2:  13795.64753ms
  Performance Run #3:  13960.317346ms
  Performance Run #4:  13530.801789ms
  Performance Run #5:  13425.006736ms

  Performance budget exceeded: 13530.801789ms (limit: 11142.857142857143ms)
```

After:

```
Single File:
  CPU Speed is 3500 with multiplier 13000000
  Performance Run #1:  3705.140804ms
  Performance Run #2:  3714.937271ms
  Performance Run #3:  3817.76126ms
  Performance Run #4:  3948.772472ms
  Performance Run #5:  3707.20636ms

  Performance budget exceeded: 3714.937271ms (limit: 3714.285714285714ms)


Multi Files (0 files):
  CPU Speed is 3500 with multiplier 39000000
  Performance Run #1:  13963.964251ms
  Performance Run #2:  13807.104047ms
  Performance Run #3:  14118.498804ms
  Performance Run #4:  14127.543946ms
  Performance Run #5:  13701.669443ms

  Performance budget exceeded: 13963.964251ms (limit: 11142.857142857143ms)
```

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,276 @@ class Config {
+     * Gets the vector of applicable configs from the hierarchy for a given file, including any overrides that apply to
+     * the specified file path. A vector is an array of config file paths, each optionally followed by one or more
+     * numbers that correspond to the indices of nested config blocks within the config file's overrides section whose
+     * glob patterns match the specified file path; e.g., the vector ['/home/john/project/app/.eslintrc', 0, 2] would
```
Is there a reason to use an array with strings and numbers here, rather than some other data structure? An array containing multiple datatypes with different meanings seems like a code smell.

Maybe an object would be better:

```js
{
  configFilePath: '/home/john/project/app/.eslintrc',
  overrideConfigs: [
    {
      rules: { /* ... */ }
    },
    {
      rules: { /* ... */ }
    }
  ]
}
```

not-an-aardvark commented on lib/config/config-file.js:
```diff
@@ -547,6 +548,43 @@ function load(filePath, applyEnvironments, relativeTo) {
+ */
+function loadCached(filePath) {
+    const cachedConfig = configCache.getConfig(filePath);
+    let config = {};
```
Initializing `config` to an empty object seems unnecessary here, because it's reassigned immediately below.

smably replied:
Fixed.

not-an-aardvark commented on package.json:
```diff
@@ -55,6 +55,7 @@
     "json-stable-stringify": "^1.0.0",
     "levn": "^0.3.0",
     "lodash": "^4.0.0",
+    "minimatch": "^3.0.0",
```
We already have a dependency on `glob`, which we use for parsing CLI options. Can you use that instead of adding another glob-parsing dependency?

smably replied:
I agree that this seems redundant, but `glob` is specifically for matching patterns against files and directories on disk. In this case, `minimatch` is being used to check a file path (which is already confirmed to exist) against a specified pattern, which `glob` is not able to do. And `glob` actually uses `minimatch` under the hood -- it just adds the filesystem traversal logic -- so the results should be the same.

smably replied:
I realized that this is not going to work: `minimatch` actually doesn't behave the same way as `glob`, in that `glob` requires the _whole_ filename to match the pattern, whereas minimatch only checks that _any_ part of the filename matches. So an override rule that's supposed to match `**/test.js` is actually matching, e.g., `src/components/Component.test.jsx`. I'm wondering whether it would make sense to enumerate the list of files matching every glob pattern in `overrides` and then find the intersection between those files and the files matched from the CLI glob pattern. The downside is that there would definitely be a performance penalty. I'll benchmark and see how it compares.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,276 @@ class Config {
+            if (value) {
+                this.cliConfig[configKey] = value;
+            }
+        }, this);
```
Nitpick: If you use an arrow function here, you won't need to add a `this` argument at the end.

smably replied:
Fixed.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,276 @@ class Config {
+            cache = configCache.getHierarchyLocalConfigs(localConfigDirectory);
+            if (cache) {
+                break;
+            }
```
I'm not sure I'm understanding this logic -- aren't all local config files in the same directory? Why does the cache need to be checked separately for each iteration of this loop?

smably replied:
Local config files can be in the current directory, but they can also be in the parent directory, or any directory above the current one in the tree.

`findLocalConfigFiles` actually finds all the config files in the filesystem above the specified directory. So `localConfigFiles` will contain config files in various directories above the current one, some of which may have already been cached. This logic stops walking up the directory tree as soon as we hit a directory that's already in the local hierarchy cache, because it already contains all the configs for that directory and its parents.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,276 @@ class Config {
-            Plugins.loadAll(this.options.plugins);
-            config = ConfigOps.merge(config, { plugins: this.options.plugins });
-        }
+        if (!configs.length && !cache && !this.useSpecificConfig) {
```
This logic is hard to follow. Would it be possible to split it into multiple functions?

Specifically, it's not clear to me how the presence of a cache changes the control flow. Intuitively, I would expect the cache to be an alternate way to populate `configs`, but in that case its effects should be isolated, and it shouldn't be necessary to refer to the cache separately afterwards.

not-an-aardvark commented on lib/config/config-cache.js:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    /**
+     * Gets a list of hierarchy-local config objects that apply to the specified directory.
```
What are "hierarchy-local config objects"?

smably replied:
Configs parsed from config files in the filesystem hierarchy that apply to files in directories below them (as opposed to, say, the user's personal config from their homedir, or a config file passed as on option on the command line). Totally willing to change the terminology if you can think of a better way to describe these.

not-an-aardvark commented on lib/config/config-ops.js:
```diff
@@ -268,5 +272,84 @@ module.exports = {
+    getConfigFromVector(vector) {
+
+        const subvector = Array.from(vector);
+        let config,
```
The `config` variable is reassigned in several different places, and it's mutated later on, which makes this code hard to follow. Could it be restructured into functions to make it easier to tell what's going on?

smably replied:
Fixed.

smably commented on lib/config.js:
```diff
@@ -212,127 +87,276 @@ class Config {
+                return;
+            }
+
+            const relativePath = (filePath || directory).substr(config.baseDirectory.length + 1);
```
Fixed.

smably commented on lib/config/config-ops.js:
```diff
@@ -268,5 +272,84 @@ module.exports = {
+            nearestCacheIndex = vector.length - 1;
+
+        // Check the merged vector config cache to try to find a merged config for the current config or a parent
+        for (; nearestCacheIndex >= 0; nearestCacheIndex--) {
```
Fixed.

smably commented:
Thanks for the comments, @ilyavolodin and @not-an-aardvark! It's been a busy week, and I haven't had a chance to address all of the issues you identified, but I've pushed my first round of fixes and responded to some of the comments. Just wanted to confirm that I haven't forgotten about this, and further fixes are coming.

not-an-aardvark commented:
@smably I see you've pushed some changes since this was last reviewed. Do you think this is ready to be reviewed again? (There's no rush -- I just want to make sure we're not accidentally deadlocked if you were waiting for a review and I was waiting for you to finish.)

smably commented:
Hey @not-an-aardvark, thanks for the nudge. I fixed most of the issues but tests seems to be failing on Windows (I'm guess something path-related) and I've been procrastinating figuring out the issue because I don't have a Windows machine available. I'll see if I can find some time in the next week to do some debugging and update the thread with the status of this work.

not-an-aardvark commented:
It looks like the issue is that you're using `Array.prototype.includes`, which only exists on Node 6+, but the build runs on Node 4+. Switching to `indexOf` should hopefully fix the problem.

smably commented:
@not-an-aardvark Thanks for the catch. I fixed and rebased. Builds look OK now! When you get a chance, I'd appreciate another look over this.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
+     * Builds a hierarchy of config objects, including the base config, all local configs from the directory tree,
+     * and a config file specified on the command line, if applicable.
+     * @param {string} directory a file in whose directory we start looking for a local config
+     * @returns {Object[]} The config objects
```
Can you update this comment to specify how the config objects are ordered in the array? (Are the config objects with higher precedence placed at the beginning of the array, or the end?)

smably replied:
Done.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
+     * Gets a list of config objects extracted from local config files that apply to the current directory, in
+     * descending order, beginning with the config that is highest in the directory tree.
+     * @param {string} directory The directory to start looking in for local config files.
+     * @returns {Object[]} The shallow local config objects, or an empty array if there are no local configs.
```
Can you update this comment to specify how the config objects are ordered in the array? (Are the config objects with higher precedence placed at the beginning of the array, or the end?)

smably replied:
Done

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
-            });
-        }
+            // Check for root flag
+            if (localConfig.root === true) {
```
Nitpick: This can just be `if (localConfig.root)` because the property is required to be a boolean if it appears at all.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
+            searched = [],
+            configs = [];
+        let rootPath,
+            cache;
```
Nitpick: Can you give this variable a more descriptive name? What is it caching?

smably replied:
Fixed.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
-        debug("Merging command line environment settings");
-        config = ConfigOps.merge(config, { env: this.env });
+            // If root flag is set, don't consider file if it is above root
+            if (rootPath && !pathIsInside(path.dirname(localConfigFile), rootPath)) {
```
If I'm understanding correctly, this loop will always traverse directories starting with the most nested subdirectory and repeatedly look in the parent directory.

Would it work to just `break;` after a file is found with `root: true`? That way, this `if` statement would be unnecessary because none of the parent directories of the file with `root: true` would get traversed.

smably replied:
Good call! This logic cleaned up nicely.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
-        this.cache[directory] = config;
+        // Merged with any cached portion
+        configs.reverse();
+        cache = cache ? cache.concat(configs) : configs;
```
Instead of doing this, would it be possible to append `cache` immediately after `configCache.getHierarchyLocalConfigs(localConfigDirectory)`? That way, `cache` wouldn't need to be referenced outside of the loop, and the function would just return `configs`.

smably replied:
Good call. I've cleaned up this logic.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
+     * Gets the vector of applicable configs and subconfigs from the hierarchy for a given file. A vector is an array of
+     * objects, each containing a config file path and an array of override indices corresponding to entries in config
+     * file's overrides section whose glob patterns match the specified file path; e.g., the vector
+     * { configFile: '/home/john/app/.eslintrc', matchingOverrides: [0, 2] } would indicate that the main project
```
Is a vector an array of objects, or a single object? Based on the top of the comment it seems like it's an array of objects, but then this line is a bit misleading because it calls a single object a vector.

smably replied:
Clarified this comment. Thanks.

not-an-aardvark commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
+     * { configFile: '/home/john/app/.eslintrc', matchingOverrides: [0, 2] } would indicate that the main project
+     * .eslintrc file and its first and third override blocks apply to the current file.
+     * @param {string} filePath The file path for which to build the hierarchy and config vector.
+     * @returns {Array<Object>} config vector applicable to the specified path
```
Would it be possible to have a vector just contain the override objects themselves rather than a filePath and indices?

smably replied:
It's an interesting idea, but I'm not sure how exactly it would work. Right now, the vector is used as a sort of unique identifier that can be hashed to fetch a merged config object for a file. The logic is: take the file's path, match it against the override patterns from all the configs above it in the hierarchy to get a vector, then hash the vector to see whether there exists a merged config for that exact combination of configs + overrides. If the vector contained the overrides themselves, we'd still need a way to figure out whether there's a merged config in the cache for that file, otherwise we'd have to do the work of merging the overrides for every file. So I don't think it would gain us anything, unless I'm misunderstanding what you're proposing.

not-an-aardvark commented on lib/config/config-cache.js:
```diff
@@ -0,0 +1,164 @@
+ * @private
+ */
+function hashVectorEntry(entry) {
+    return [].concat(entry.filePath, entry.matchingOverrides).join(VECTOR_SEP);
```
I think this could create hash collisions if a file path contains a comma. For example, someone could name a file "foo.js,0".

smably replied:
Good catch. I've changed it to use the JSON-serialized representation of the vector as the hash.

ilyavolodin replied:
Can we use Symbol for this instead? Or we can't recreated Symbol for access?

smably replied:
I don't think we could use a symbol here, because we're essentially doing a deep equality check on the two vectors. We'd have to figure out the correct mapping from a vector to a symbol, which is really the same problem as directly testing the equality of the vectors.

not-an-aardvark commented on docs/user-guide/configuring.md:
```diff
@@ -712,6 +712,61 @@ module.exports = {
 
+## Configuration Based on Glob Patterns
+
+Sometimes a more fine-controlled configuration is necessary, for example if the configuration for files within the same directory has to be different. Therefore you can provide configurations under the `overrides` key that will only apply to files that match specific glob patterns, using the same format you would pass on the command line (e.g., `app/**/*.test.js`).
```
I think you'll need to update the [config schema](https://github.com/eslint/eslint/blob/c70b0eda8794e552bcbfccbc0d5f46dfb4ea7783/conf/config-schema.json) because ESLint now throws f it encounters an unknown top-level property.

smably replied:
Fixed.

smably commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
-        const useConfig = options.configFile;
-
-        this.options = options;
+        let useConfig = options.configFile;
```
Fixed -- thanks!

smably commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
+            Plugins.loadAll(this.options.plugins);
         }
+
+        // Empty values in configs don't merge properly
```
I tried allowing empty values in the config and a bunch of tests failed, so I'm pretty sure this is still necessary.

smably commented on lib/config.js:
```diff
@@ -212,127 +87,277 @@ class Config {
+            if (value) {
+                this.cliConfig[configKey] = value;
+            }
+        });
```
See above. Passing `plugins` as `undefined` threw a hard error, and setting other keys to `undefined` made tests fail.

smably commented:
@not-an-aardvark Thanks for the comments! Fixed most of the issues you flagged, plus left a few comments in response.

not-an-aardvark commented on tests/fixtures/config-hierarchy/overrides/.eslintrc:
```diff
@@ -0,0 +1,38 @@
+        {
+            "files": "foo.js",
+            "rules": {
+                "quotes": [2, "single"]
```
Nitpick: Can you use different rules for the different overrides tests here rather than using `quotes` for all of them? If there is a bug where some overrides get applied that shouldn't be, I want to make sure the bug is caught by the tests.

smably replied:
Done. Test configs are much more realistic now!

not-an-aardvark commented on tests/lib/config.js:
```diff
@@ -1055,6 +1055,205 @@ describe("Config", () => {
+                assertConfigsEqual(actual, expected);
+            });
+
+            it("should merge all local configs (override and non-override) before non-local configs", () => {
```
Can you add a test where a config is in `overrides` in a parent directory, but not in `overrides` in a child directory?

For example:

```yaml
# root .eslintrc.yml

overrides:
  files: 'child-dir/**/*.js'
  config:
    rules:
      semi: [error, always]
```

```yml
# child-dir/.eslintrc.yml

rules:
  semi: [error, never]
```

not-an-aardvark commented on tests/lib/config.js:
```diff
@@ -1055,6 +1055,205 @@ describe("Config", () => {
+                    cwd: getFakeFixturePath("overrides"),
+                    baseConfig: {
+                        overrides: [{
+                            files: ["one/**/*", "*.js"],
```
It looks like the file used in this test would match both of the globs in this array. Can you add a test where it matches some globs but does not match other globs?

smably replied:
Done.

smably commented on lib/config.js:
```diff
@@ -184,24 +51,31 @@ function getLocalConfig(thisConfig, directory) {
+        this.baseConfig.filePath = "";
+        this.baseConfig.baseDirectory = this.options.cwd;
+
+        configCache.reset();
```
Refactored to create a new `ConfigCache` in the `Config` constructor rather than using a singleton.

smably commented on lib/config.js:
```diff
@@ -212,127 +86,276 @@ class Config {
+        debug(`Constructing config file hierarchy for ${directory}`);
+
+        // Step 1: Always include baseConfig
+        configs = [this.baseConfig];
```
Fixed.

smably commented on lib/config.js:
```diff
@@ -212,127 +86,276 @@ class Config {
+        const directory = filePath ? path.dirname(filePath) : this.options.cwd,
+            vector = [];
+
+        this.getConfigHierarchy(directory).forEach(config => {
```
Fixed.

smably commented on lib/config/config-cache.js:
```diff
@@ -0,0 +1,148 @@
+ * @private
+ */
+function hash(vector) {
+    if (!vector) {
```
Fixed.

smably commented on lib/config/config-file.js:
```diff
@@ -571,6 +572,42 @@ function load(filePath, applyEnvironments, relativeTo) {
+ * @returns {Object} the parsed config object (empty object if there was a parse error)
+ * @private
+ */
+function loadCached(filePath) {
```
Renamed `loadCached` to `load`.

smably commented on lib/config/config-ops.js:
```diff
@@ -268,5 +273,102 @@ module.exports = {
+     */
+    pathMatchesGlobs(filePath, patterns) {
+        const validPatterns = [].concat(patterns).filter(
+            pattern => !path.isAbsolute(pattern) && pattern.indexOf("..") === -1
```
Fixed.

smably replied:
Good call. Made it throw in either of those cases.

smably commented on tests/lib/config.js:
```diff
@@ -1055,6 +1055,205 @@ describe("Config", () => {
+
+            /**
+             * Returns the path inside of the fixture directory.
+             * @returns {string} The path inside the fixture directory.
```
Fixed (though there are a bunch of other functions in this file that use the same pattern that could probably use a refactor).

smably commented:
@not-an-aardvark Hey, I finally got around to fixing this up. Sorry that took a while -- had to do a very painful rebase that also broke a bunch of tests in the process. Unfortunately, I squashed a bunch of the commits to try to make the merge conflict fixing easier, so I apologize if the commits aren't all cleanly separated. Hopefully fixed all the issues raised in the last round of review, but there's been enough refactoring that this may deserve another full check.

not-an-aardvark commented:
No worries. Sorry about that, there has been a lot of churn in the codebase in the past few months.

not-an-aardvark commented on docs/user-guide/configuring.md:
```diff
@@ -713,6 +713,61 @@ module.exports = {
+
+### How it works
+
+* Glob pattern overrides can only be configured within `.eslintrc` files.
```
This line is a bit confusing, since overrides can be configured in any config files (such as `.eslintrc.json`, `.eslintrc.yml`, or the `eslintConfig` property of a `package.json` file), not just `.eslintrc` files.

On a related note, since `.eslintrc` is deprecated in favor of the versions with file extensions, maybe we should use file extensions in the examples.

smably replied:
Fixed docs to remove references to `.eslintrc`.

not-an-aardvark commented on docs/user-guide/configuring.md:
```diff
@@ -713,6 +713,61 @@ module.exports = {
+
+* Glob pattern overrides can only be configured within `.eslintrc` files.
+* The patterns are applied against the file path relative to the directory of the `.eslintrc` file. For example, if your `.eslintrc` file has the path `/Users/john/workspace/any-project/.eslintrc` and the file you want to lint has the path `/Users/john/workspace/any-project/lib/util.js`, then the pattern provided in `.eslintrc` will be executed against the relative path `lib/util.js`.
+* Glob pattern overrides have higher precedence than the regular configuration in the same `.eslintrc` file.
```
Can you clarify the precedence between two overrides blocks in the same file? Does the later override take precedence?

```json
{
  "overrides": [
    {
      "files": ["foo.js"],
      "config": {
        "rules": {
          "semi": ["error", "always"]
        }
      }
    },
    {
      "files": ["*.js"],
      "config": {
        "rules": {
          "semi": ["error", "never"]
        }
      }
    }
  ]
}
```

More specifically, what would be the config for `foo.js` in the above example?

smably replied:
Done. (Later overrides take precedence.)

not-an-aardvark commented on lib/config.js:
```diff
@@ -209,132 +90,275 @@ class Config {
+                config = path.resolve(this.options.cwd, config);
             }
+
+            this.useSpecificConfig = ConfigFile.load(config, this);
```
Nitpick: Can you rename this property to something like `specificConfig` instead to clarify that it's a config object and not a boolean?

smably replied:
Fixed.

not-an-aardvark commented on docs/user-guide/configuring.md:
```diff
@@ -713,6 +713,61 @@ module.exports = {
+    "quotes": [ 2, "double" ]
+  },
+
+  "overrides": [
```
What happens if a config has nested overrides?

```json
{
  "overrides": [
    {
      "files": ["*.js"],
      "config": {
        "rules": {
          "semi": ["error", "never"]
        },
        "overrides": [
          {
            "files": ["bar.js"],
            "config": {
              "rules": {
                "semi": ["error", "always"]
              }
            }
          }
        ]
      }
    }
  ]
}
```

Would this break the "vector" caching strategy? If it's not supported, we should probably clarify that in the docs and make it a fatal error to avoid confusion.

smably replied:
Nested overrides are not allowed. I clarified the docs to make this more explicit and added schema validation so this will throw an error right away.

not-an-aardvark commented on lib/config/config-ops.js:
```diff
@@ -268,5 +273,108 @@ module.exports = {
+            }
+        });
+
+        return patternList.length > 0 && patternList.every(
```
This might change if we modify the semantics to use the union of the globs instead of the intersection, but it seems like it might be confusing to have a special case for when the `files` array is empty.

smably replied:
Good catch. I can't imagine any valid use case for an empty `files` array, so I added a length check in the schema and removed this special case logic.

not-an-aardvark commented on lib/config/config-file.js:
```diff
@@ -547,27 +543,60 @@ function load(filePath, configContext, applyEnvironments, relativeTo) {
+    const config = loadFromDisk(resolvedPath, configContext);
+
+    if (config) {
+        config.filePath = resolvedPath.filePath;
```
Will adding properties to the config object affect the result of `eslint --print-config someFile.js`? Ideally, it would be best to avoid exposing internal properties to the user.

One way to do this could be to have a map of filePath -> config rather than adding a property.

smably replied:
I checked, and `eslint --print-config` doesn't print these internal properties, so I think this is OK.

smably commented on docs/user-guide/configuring.md:
```diff
@@ -713,6 +713,61 @@ module.exports = {
+* The patterns are applied against the file path relative to the directory of the `.eslintrc` file. For example, if your `.eslintrc` file has the path `/Users/john/workspace/any-project/.eslintrc` and the file you want to lint has the path `/Users/john/workspace/any-project/lib/util.js`, then the pattern provided in `.eslintrc` will be executed against the relative path `lib/util.js`.
+* Glob pattern overrides have higher precedence than the regular configuration in the same `.eslintrc` file.
+* A glob specific configuration works almost the same as the regular configuration in your `.eslintrc`, except that you can't define nested glob based configurations or use `extends`.
+* Multiple glob patterns can be provided within a single override block. The associated configuration will be applied to the intersection of the sets of files matched by each of the patterns. In other words, a file must match all the supplied patterns for the configuration to apply.
```
My reasoning was that it's easy to construct a union with a glob pattern, e.g., `{foo,bar}/**/*.js`, but I don't think it's possible to do complex filtering with negation within a single glob pattern.

E.g., to match all files except tests within a specific directory, you could do something like:
```json
{
  "overrides": [
    {
      "files": ["foo/**/*.js", "!foo/**/*.test.js"]
    }
  ]
}
```

This seems like a valuable use case that should maybe be called out explicitly in the docs.

not-an-aardvark replied:
I see, that's a good point. I hadn't realized that the `{foo,bar}` syntax exists for globs.

I'm concerned that this is semantically confusing, though -- I suspect a lot of people would expect a config labelled `files: ["foo.js", "bar.js"]` to apply to the files `foo.js` and `bar.js`, but as it currently stands the config will silently not match anything.

smably replied:
Yeah, that is fair. What do you think about renaming `files` to `matchAll` or something along those lines?

randycoulman replied:
I don't wanna bikeshed this too much because I really want to see this merged!  But the intersection behavior would confuse me based on my experience with other configuration systems; I would expect it to be a union behavior.  I don't think the `{foo,bar}` glob syntax is very well known.  Maybe if it's well-documented here, that's enough?

Here's another option, though.  What if `files` as an array had union semantics, but it could also optionally be an object with `includes` and `excludes` keys.  An array would be equivalent to `files: { includes: [...] }`.

With this, `["foo.js", "bar.js"]` would work as @not-an-aardvark expects, and @smably's use case could be handled as: `files: { includes: ["foo/**/*.js"], excludes: ["foo/**/*.test.js"] }`.  `excludes` could also be `except` to better communicate that it comes after the inclusion.

smably replied:
Yeah, I wondered about adding an option with explicit inclusion/exclusion keys. My only concern is the complexity, since that will make three ways to specify globs (string, array, or object). I lean toward keeping it simple and renaming the property, but I don't feel strongly about it. Any thoughts @not-an-aardvark?

not-an-aardvark replied:
> What do you think about renaming files to matchAll or something along those lines?

> What if files as an array had union semantics, but it could also optionally be an object with `includes` and `excludes` keys. An array would be equivalent to files: `{ includes: [...] }`

Both of those ideas seem like they could be improvements. If we keep `files` with union semantics, one alternative would be to also have a separate `excludedFiles` property rather than allowing `files` to be an object. That way we wouldn't have to allow `files` to be two different types.

smably replied:
Aha, I like that last option, a separate `excludedFiles` property. If we have consensus, I'm happy to go ahead with that.

not-an-aardvark replied:
Sounds good to me.

smably replied:
Done.

smably commented:
@not-an-aardvark I think this one's ready for another review. Almost there, I hope!

not-an-aardvark commented on docs/user-guide/configuring.md:
```diff
@@ -713,6 +713,61 @@ module.exports = {
+    {
+      "files": [ "bin/*.js", "lib/*.js" ],
+      "excludedFiles": "*.test.js",
+      "rules": {
```
Was it intentional that this commit removed the `config` property and put the rest of the config in the the `overrides` object? I could go either way on this, but it seems like there would be a benefit to making it clear that `files` and `excludedFiles` aren't regular config keys.

smably replied:
I'm pretty sure this was just a mistake in the docs, because this proposal has put the globs at the same level as the other config properties since the beginning (see #3611). I don't feel strongly either way -- on the one hand, it might be nice to separate the config from the override globs, but on the other hand, it would be yet one more level of nesting. Unless anyone has a strong preference, I guess I'll leave it as is?

not-an-aardvark commented on conf/config-schema.js:
```diff
@@ -0,0 +1,62 @@
+    baseConfigProperties,
+    {
+        files: globList,
+        excludedFiles: globList
```
This isn't a blocker, but it seems like it might be useful to allow empty excludedFiles arrays -- I know some people like to explicitly state default options in their config (e.g. in case they want to exclude files later, they won't have to look up the name again). At the moment, `excludedFiles` has `minItems: 1`.

smably replied:
Good point. Fixed.

not-an-aardvark commented:
Waiting another day to merge this in case anyone else on @eslint/eslint-team wants to review it.

smably commented:
🎉  Thank you for the guidance and very helpful review comments! This is my first open source contribution (of code, at least) and at first I was a little worried that I'd bitten off more than I could chew. But I think it came together quite nicely.

ilyavolodin commented:
Awesome job @smably This is a very high profile feature that users have been asking for a while, so great job at taking on it!

mischah commented:
This is great. @smably Thanks for investing your time into that ❤️

Moeriki commented:
I noticed I couldn't use `extends: []` in `overrides: []`.

Eg.

```js
{
  overrides: [{
    files: ['test/**'],
    extends: ['my-config/mocha']
  }],
}
```

It that an intentional limitation?

platinumazure commented:
Ping @smably @not-an-aardvark, see [this comment](https://github.com/eslint/eslint/pull/8081#issuecomment-311035672).

If we're intentionally not supporting `extends` in override configs, I think we might want to see if we can enhance that (as a separate issue). My suggested approach would be to prepend the override's `extends` array to the parent config's `extends` array and removing duplicate entries that occur later in the resulting array, but there might be a better way to do that.

smably commented:
@platinumazure Yup, intentionally unsupported (and called out in the docs). It wasn't in the PR I inherited, and I didn't want to add any more scope to what was already a big change. I agree that adding support for this would make a worthy follow-up ticket, and I would be happy to work on the implementation if nobody else has already started.

platinumazure commented:
@smably Awesome, thanks for clarifying.

I'll create a new issue.

CrabDude commented:
I am beyond impressed y'all saw this through to fruition, and honestly a little shocked. Awesome.

❤️ ❤️  @smably @not-an-aardvark @ilyavolodin

CrabDude commented:
Does anyone have final performance numbers on a codebase of significant size?

kaicataldo commented:
I just used this feature on a large project at work - really great stuff :heart:

smably commented:
@CrabDude Thanks for getting this going and figuring out the logic! You definitely deserve a lot of the credit for this feature... no way I would have been able to get all that caching logic working on my own. ❤️
